lib/shopify_transporter/pipeline/magento/product/top_level_variant_attributes.rb

```diff
@@ -8,18 +8,15 @@ module Magento
       module Product
         class TopLevelVariantAttributes < Pipeline::Stage
           def convert(hash, record)
-            return unless input_applies?(hash)
-            simple_product_in_magento_format = record['variants'].select do |product|
+            return {} unless record.key?('variants')
+            simple_product_in_magento_format = record['variants'].find do |product|
```

Comment:
this is the most confusing part of the PR. it turns out that you need to do this instead of just directly passing `hash` to the accumulator, because the accumulator modifies what is passed in, so you specifically need to pass in the copy of the `hash` that already exists in `record` thanks to the `VariantAttributes` pipeline stage.

Reply:
if you're confused about this, well, me too. we should really really rewrite this in a way that this doesn't happen

Reply:
This is the sort of code that reminds me how annoying mutation is. For example why doesn't this work?

```
          def convert(hash, record)
            simple_product = corresponding_variant_in_record(hash, record)
            return {} unless simple_product.present?
            accumulator = TopLevelVariantAttributesAccumulator.new(simple_product)
            accumulator.accumulate(hash)
          end

          def corresponding_variant_in_record(hash, record)
            return unless record.key?('variants')
            simple_product_in_magento_format = record['variants'].find do |product|
              product['product_id'] == hash['product_id']
            end
            return unless simple_product_in_magento_format.present?
          end
```

It seems like this should be equivalent to the code as written. However, it causes all the tests to fail.

I'm assuming this has to do with Ruby's pass-reference-by-value stuff or whatever, but I don't really know. At this point I want to just touch it as little as possible and make the fix I need to make without breaking anything.

Reply:
And the fact that I feel that way is indicative of tech debt! If your code is so confusing that you feel afraid to touch it... that's real bad.